src/gflownet/trainer.py

```diff
@@ -199,103 +210,95 @@ def _wrap_for_mp(self, obj, send_to_device=False):
         if send_to_device:
             obj.to(self.device)
         if self.cfg.num_workers > 0 and obj is not None:
-            wapper = mp_object_wrapper(
+            wrapper = mp_object_wrapper(
                 obj,
                 self.cfg.num_workers,
                 cast_types=(gd.Batch, GraphActionCategorical, SeqBatch),
                 pickle_messages=self.cfg.pickle_mp_messages,
             )
-            self.to_terminate.append(wapper.terminate)
-            return wapper.placeholder, torch.device("cpu")
+            self.to_terminate.append(wrapper.terminate)
+            return wrapper.placeholder, torch.device("cpu")
         else:
             return obj, self.device
 
     def build_callbacks(self):
         return {}
 
-    def build_training_data_loader(self) -> DataLoader:
-        model, dev = self._wrap_for_mp(self.sampling_model, send_to_device=True)
-        replay_buffer, _ = self._wrap_for_mp(self.replay_buffer, send_to_device=False)
-        iterator = SamplingIterator(
-            self.training_data,
-            model,
-            self.ctx,
-            self.algo,
-            self.task,
-            dev,
-            batch_size=self.cfg.algo.global_batch_size,
-            illegal_action_logreward=self.cfg.algo.illegal_action_logreward,
-            replay_buffer=replay_buffer,
-            ratio=self.cfg.algo.offline_ratio,
-            log_dir=str(pathlib.Path(self.cfg.log_dir) / "train"),
-            random_action_prob=self.cfg.algo.train_random_action_prob,
-            det_after=self.cfg.algo.train_det_after,
-            hindsight_ratio=self.cfg.replay.hindsight_ratio,
-        )
-        for hook in self.sampling_hooks:
-            iterator.add_log_hook(hook)
+    def _make_data_loader(self, src):
         return torch.utils.data.DataLoader(
-            iterator,
+            src,
             batch_size=None,
             num_workers=self.cfg.num_workers,
             persistent_workers=self.cfg.num_workers > 0,
             prefetch_factor=1 if self.cfg.num_workers else None,
         )
 
+    def build_training_data_loader(self) -> DataLoader:
+        # Since the model may be used by a worker in a different process, we need to wrap it.
+        # The device `dev` returned here is the device that the worker will use to interact with the model;
+        # normally, if the main process has the model on 'cuda', this will simply be 'cpu' (since workers
+        # don't have CUDA access).
+        # See implementation_nodes.md for more details.
+        model, dev = self._wrap_for_mp(self.sampling_model, send_to_device=True)
+        replay_buffer, _ = self._wrap_for_mp(self.replay_buffer, send_to_device=False)
+
+        n_drawn = int(self.cfg.algo.global_batch_size * (1 - self.cfg.algo.offline_ratio))
+        n_replayed = n_drawn if self.cfg.replay.batch_size is None else self.cfg.replay.batch_size
```

Comment:
Hmm, that's a good point. I wonder if it wouldn't be even simpler to just have these directly as four hyperparameters:
- `n_from_dataset`
- `n_from_policy`
- `n_replayed`
- `n_keep_in_batch`, Optional, and defaults to `n_from_policy` if unspecified (this would match the current behavior, but we could change it)

and then, if the users want to think of these as ratios of some global batch size, that's up to them.

I did all this calculus to not change the default behavior, but maybe it's better to "break" things (well, not really break as long as we change the tasks' `default_hps` accordingly) and also make them clearer.

Reply:
Yep I like that, I think that would be the cleanest way to go

Reply:
Actually maybe just:

- `n_from_dataset`
- `n_from_policy`
- `n_from_replay`

We don't really need `n_keep_in_batch` right? Everything `from_policy` goes into the buffer and stays in the batch. If `n_from_replay > n_from_replay` make sure that `replay.warmup >= n_from_replay`.

Reply:
In our current implementation we override the policy samples with replay samples, i.e. `n_keep_in_batch` is 0. That's standard in RL, but afaik not standard in GFNs. Happy to drop `n_keep_in_batch` and change the default though (consider what GFN people seem to do).

Reply:
Yeah I agree, I think most of the time people would exclusively use `from_policy` or `from_replay` or `from_dataset and from_policy`, but having the 3 settable in an independent way makes it easier to study the effect of online v.s. off-policy v.s. offline while keeping the config crystal clear.

Reply:
Following up on our discussion I ended going with:
- `algo.num_from_policy`
- `algo.num_from_dataset`
- `replay.num_from_replay`, Defaults to `algo.num_from_policy`, (i.e. 50/50)
- `replay.num_new_samples`, Defaults to `algo.num_from_policy`, if smaller than `num_from_policy` then not all on-policy samples will be added to the replay. If larger than `num_from_policy` then the training batch will not contain all the new samples, but the buffer will. For example, if one wishes to sample `N` samples every step but only add them to the buffer and not make them part of the training batch, then one should set `replay.num_new_samples=N` and `algo.num_from_policy`=0.